geniza/entities/models.py

```diff
@@ -146,6 +151,168 @@ def __str__(self):
         except Name.DoesNotExist:
             return str(self.names.first() or super().__str__())
 
+    def merge_with(self, merge_people, user=None):
+        """Merge the specified people into this one. Combines all metadata
+        into this person and creates a log entry documenting the merge.
+
+        Closely adapted from :class:`Document` merge."""
+
+        # if user is not specified, log entry will be associated with script
+        if user is None:
+            user = User.objects.get(username=settings.SCRIPT_USERNAME)
+
+        # language codes are needed to merge description, which is translated
+        language_codes = [lang_code for lang_code, lang_name in settings.LANGUAGES]
+
+        # handle translated description: create a dict of descriptions
+        # per supported language to aggregate and merge
+        description_chunks = {
+            lang_code: [getattr(self, "description_%s" % lang_code)]
+            for lang_code in language_codes
+        }
+
+        # collect names as strings (since that's the important part for comparison)
+        self_names = [str(name) for name in self.names.all()]
+
+        for person in merge_people:
+            # ensure any has_page overrides are respected
+            if person.has_page and not self.has_page:
+                self.has_page = True
+
+            # ensure missing role and gender are migrated
+            if person.role and not self.role:
+                self.role = person.role
+            if self.gender == Person.UNKNOWN and person.gender != Person.UNKNOWN:
+                self.gender = person.gender
+
+            # combine log entries (before name, since name used in log entries)
+            self._merge_logentries(person)
+
+            # combine names
+            for name in person.names.all():
+                if str(name) not in self_names:
+                    # if not duplicated, make non-primary and add
+                    name.primary = False
+                    name.save()
+                    self.names.add(name)
+
+            # add description if set and not duplicated
+            # for all supported languages
+            for lang_code in language_codes:
+                description_field = "description_%s" % lang_code
+                person_description = getattr(person, description_field)
+                current_description = getattr(self, description_field)
+                if person_description and person_description not in current_description:
+                    description_chunks[lang_code].append(
+                        "Description from merged entry:\n%s" % (person_description,)
+                    )
+
+            # combine person-person relationships
```

Comment:
complicated - a little hard to follow!

Reply:
Hopefully a little easier now, though I have some regrets about the naming scheme of "from" and "to"... it isn't the best, but maybe not worth changing at this point.

Reply:
curious what you would you use instead - have you thought of something more semantic?  agree it probably isn't worth changing

I've wondered if these relationships would be easier to express as RDF with subject, predicate, object (although that opens up a different can of worms...)

Reply:
I didn't have anything in mind, and can't think of anything off the top of my head—which contributes to the inertia to not change it! 😆